course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,447 @@
+
+
+<center>
+    <img src="images/exploring-data-structures/concat.png" 
+    width=700 style="margin:auto"/>
+    <p style="text-align: center">
+       Ways to stack data
+    </p>
+    <br/>
+</center>
+
+
+
+To illustrate, let’s fetch two simple dataframes. Each contains the average scores for three subjects by year for two separate schools.
+
+```{python}
+school1_url  = "https://raw.githubusercontent.com/bcgov/ds-intro-to-python/main/data/school1.csv"
+school2_url  = "https://raw.githubusercontent.com/bcgov/ds-intro-to-python/main/data/school2.csv"
+school1_df = pd.read_csv(school1_url)
+school2_df = pd.read_csv(school2_url)
+```
+
+Let's take a quick look at the two dataframes.
+
+```{python}
+school1_df
+```
+
+```{python}
+school2_df
+```
+
+We call `.concat()` and pass in the objects that we want to stack vertically. This is a similiar operation to `union` in `SQL`.
+
+```{python}
+vertical_stack_df = pd.concat([school1_df, school2_df])
+vertical_stack_df
+```
+
+It is also possible to stack the data horizontally. Here it is necessary to specify the columnar axis (axis=1) as the default setting is for rows (axis=0).
+
+```{python}
+horizontal_stack = pd.concat([school1_df, school2_df], axis=1)
+horizontal_stack
+```
+
+This introduction only scratches the surface of how to leverage this way of joining datasets together. But it can be a powerful tool in the toolkit for the right use case. More detail found here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html
+
+## Grouping and summarizing
+
+Sometimes of course you would prefer to group rows together for the purpose of summarizing them in various ways.
+
+We can accomplish this using the `.groupby()` method. A `.groupby()` operation involves some combination of splitting the object, applying a function, and combining the results.  It is used together with one or more aggregation functions:
+
+* `count()`: Total number of items
+* `first()`, `last()`: First and last item
+* `mean()`, `median()`: Mean and median
+* `min()`, `max()`: Minimum and maximum
+* `std()`, `var()`: Standard deviation and variance
+* `mad()`: Mean absolute deviation
+* `prod()`: Product of all items
+* `sum()`: Sum of all items
+
+With our data, let's use these tools to get population mean, min and max by continent. Here's how it works. 
+
+First we place "continent" in the `.groupby()` function, then we pass the aggregation functions "mean", "min", and "max" as a dictionary within the `.agg()` function. This dictionary takes the column that we are aggregating - in this case life expectancy - as a key and the aggregation functions as its value.
+
+```{python}
+grouped_single = df.groupby('continent').agg({'lifeExp': ['mean', 'min', 'max']})
+grouped_single.columns = ['lifeExp_mean', 'lifeExp_min', 'lifeExp_max']
```

Comment:
worth explaining why you have to do this column rename? Would it also be worth explaining what `.reset_index()` does after a groupby?

Reply:
Agreed - the two steps here needs some unpacking - worth showing what the df looks like after each step. And (since I recently learned about it), I think `reset_index()` is important - I think getting those grouping variables back as columns is an important step.

Reply:
Good points both. Will expand.